frame/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,730 @@
+
+            IntegratedDApps::<T>::try_mutate(
+                &smart_contract,
+                |maybe_dapp_info| -> DispatchResult {
+                    let dapp_info = maybe_dapp_info
+                        .as_mut()
+                        .ok_or(Error::<T>::ContractNotFound)?;
+
+                    // If manager origin, `None`, no need to check if caller is the owner.
+                    if let Some(caller) = origin {
+                        ensure!(dapp_info.owner == caller, Error::<T>::OriginNotOwner);
+                    }
+
+                    ensure!(
+                        dapp_info.state == DAppState::Registered,
+                        Error::<T>::NotOperatedDApp
+                    );
+
+                    dapp_info.state = DAppState::Unregistered(current_era);
+
+                    Ok(().into())
+                },
+            )?;
+
+            // TODO: might require some modification later on, like additional checks to ensure contract can be unregistered.
+
+            Self::deposit_event(Event::<T>::DAppUnregistered {
+                smart_contract,
+                era: current_era,
+            });
+
+            Ok(().into())
+        }
+
+        /// Locks additional funds into dApp staking.
+        ///
+        /// In case caller account doesn't have sufficient balance to cover the specified amount, everything is locked.
+        /// After adjustment, lock amount must be greater than zero and in total must be equal or greater than the minimum locked amount.
+        ///
+        /// It is possible for call to fail due to caller account already having too many locked balance chunks in storage. To solve this,
+        /// caller should claim pending rewards, before retrying to lock additional funds.
+        #[pallet::call_index(4)]
+        #[pallet::weight(Weight::zero())]
+        pub fn lock(
+            origin: OriginFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            Self::ensure_pallet_enabled()?;
+            let caller = ensure_signed(origin)?;
+
+            let state = ActiveProtocolState::<T>::get();
+            let mut ledger = Ledger::<T>::get(&caller);
+
+            // Calculate & check amount available for locking
+            let available_balance =
+                T::Currency::free_balance(&caller).saturating_sub(ledger.locked_amount());
+            let amount_to_lock = available_balance.min(amount);
+            ensure!(!amount_to_lock.is_zero(), Error::<T>::ZeroAmount);
+
+            // Only lock for the next era onwards.
+            let lock_era = state.era.saturating_add(1);
+            ledger
+                .add_lock_amount(amount_to_lock, lock_era)
+                .map_err(|_| Error::<T>::TooManyLockedBalanceChunks)?;
+            ensure!(
+                ledger.locked_amount().saturating_add(amount_to_lock)
+                    > T::MinimumLockedAmount::get(),
```

Comment:
```suggestion
                    >= T::MinimumLockedAmount::get(),
```

A minimum amout should be valid right?

Reply:
Right, also discovered this via UT 🙂